### PR TITLE
Modify RazorPage to use DiagnosticSource for page instrumentation \ BrowserLink.

### LINKED ISSUE #3281: Consider a design where BrowserLink \ page instrumentation uses EventNotification to listen to Razor page instrumentation calls.
This would simplify the design around the instrumentation we perform in RazorPage.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3281

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak

File: src/Microsoft.AspNet.Mvc.Razor/RazorView.cs [hunk 1]
```diff
@@ -42,21 +41,20 @@ public class RazorView : IView
...
             _htmlEncoder = htmlEncoder;
             IsPartial = isPartial;
         }
 
         /// <inheritdoc />
-        public string Path
-        {
-            get { return RazorPage.Path; }
-        }
+        public string Path => RazorPage.Path;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorView.cs]: 5 #PranavPoints awarded.

File: src/Microsoft.AspNet.Mvc.Razor/IRazorPage.cs [hunk 2]
```diff
@@ -52,9 +52,9 @@ public interface IRazorPage
         bool IsPartial { get; set; }
 
         /// <summary>
-        /// Gets or sets a <see cref="IPageExecutionContext"/> instance used to instrument the page execution.
+        /// Gets or sets a <see cref="DiagnosticSource"/> instance used to instrument the page execution.
         /// </summary>
-        IPageExecutionContext PageExecutionContext { get; set; }
+        DiagnosticSource DiagnosticSource { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorPage.cs]: This is a service, so it's a bit odd to pass it around this way.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 3]
```diff
@@ -1063,12 +1062,37 @@ public void EnsureRenderedBodyOrSections()
...
+            if (DiagnosticSource != null && DiagnosticSource.IsEnabled(EndContextEvent))
+            {
+                DiagnosticSource.Write(
+                    EndContextEvent,
+                    new
+                    {
+                        httpContext = Context,
+                        path = Path,
+                        isPartial = IsPartial,
+                    });
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: This seems 1000% nicer than what we did before!

File: src/Microsoft.AspNet.Mvc.Razor/IRazorPage.cs [hunk 4]
```diff
@@ -3,10 +3,10 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.IO;
 using System.Threading.Tasks;
 using Microsoft.AspNet.Mvc.Rendering;
-using Microsoft.AspNet.PageExecutionInstrumentation;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorPage.cs]: Do we need the `Microsoft.AspNet.PageExecutionInstrumentation.Interfaces` project anymore? Looks like you've removed all `IBufferedTextWriter`, `IPageExecutionContext`, and `IPageExecutionListenerFeature` references.

**Comment by dougbu** [comment]
What about the functional tests e.g. the `RazorPageExecutionInstrumentationWebSite` site depending on old way?

Similarly do you have a reaction PR coming for the BrowserLink repo?

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 5]
```diff
@@ -1063,12 +1062,37 @@ public void EnsureRenderedBodyOrSections()
 
         public void BeginContext(int position, int length, bool isLiteral)
         {
-            PageExecutionContext?.BeginContext(position, length, isLiteral);
+            const string BeginContextEvent = "Microsoft.AspNet.Mvc.Razor.BeginInstrumentationContext";
+            if (DiagnosticSource != null && DiagnosticSource.IsEnabled(BeginContextEvent))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: How can `DiagnosticSource` be `null`?

**Comment by dougbu** [comment]
:watch:

**Comment by pranavkm** [comment]
Closing this for now. The plan's to keep the existing interface calls in place and to supplement it with DiagnosticSource. Once BrowserLink confirms it's able to consume this we'll remove the `IPageExecutionInstrumentation` pieces.

**Pull request closed without merging**